src/mds/MDCache.cc

```diff
@@ -3555,6 +3555,82 @@ void MDCache::remove_inode_recursive(CInode *in)
   remove_inode(in);
 }
 
+/**
+ * For all inodes, dirs, dentries below an inode, compose expiry
+ * messages.  This is used when giving up all replicas of entities
+ * for an MDS peer in the 'stopping' state, such that the peer can
+ * empty its cache and finish shutting down.
+ *
+ * We have to make sure we're only expiring un-referenced items to
+ * avoid interfering with ongoing stray-movement (we can't distinguish
+ * between the "moving my strays" and "waiting for my cache to empty"
+ * phases within 'stopping')
+ *
+ * @return false if we completed cleanly, true if caller should stop
+ *         expiring because we hit something with refs.
+ */
+bool MDCache::expire_recursive(
+  CInode *in,
+  map<mds_rank_t, MCacheExpire*>& expiremap,
+  CDir *subtree)
+{
+  assert(!in->is_auth());
+
+  dout(10) << __func__ << ":" << *in << dendl;
+
+  mds_rank_t owner = subtree->dir_auth.first;
+  if (expiremap.count(owner) == 0)  {
+    expiremap[owner] = new MCacheExpire(mds->get_nodeid());
+  }
+  MCacheExpire *expire_msg = expiremap[owner];
+
+  list<CDir*> ls;
+  in->get_dirfrags(ls);
+
+  int expected_refs =
+    (int)in->is_dirty() + (int)in->is_dirty_parent() + ls.size();
```

Comment:
Hmm, I think my head was in the stray stuff when writing this check.  These should always be non-auth inodes (replicas of a foreign mdsdir), so they should never be dirty.  I'm removing the permissive condition for dirtyness.